pkg/math/oracle/median.go

```diff
@@ -33,279 +41,170 @@ func NewMedianAggregator(logger *zap.Logger, cfg mmtypes.MarketMap) (*MedianAggr
-	medians types.TickerPrices,
 ) []*big.Int {
-	convertedPrices := make([]*big.Int, 0)
+	m.logger.Debug("calculating converted prices", zap.String("ticker", target.String()))
+	if len(paths.Paths) == 0 {
+		m.logger.Error(
+			"no conversion paths",
+			zap.String("ticker", target.String()),
+		)
 
+		return nil
+	}
+
+	convertedPrices := make([]*big.Int, 0)
 	for _, path := range paths.Paths {
+		// Ensure that the number of operations is valid.
+		if len(path.Operations) == 0 || len(path.Operations) > MaxConversionOperations {
+			m.logger.Error(
+				"invalid number of operations",
+				zap.String("ticker", target.String()),
+				zap.Any("path", path),
+			)
+
+			continue
+		}
+
 		// Calculate the converted price.
-		convertedPrice, err := m.CalculateConvertedPrice(ticker, path, medians)
+		adjustedPrice, err := m.CalculateAdjustedPrice(target, path.Operations)
 		if err != nil {
 			m.logger.Debug(
 				"failed to calculate converted price",
 				zap.Error(err),
-				zap.String("ticker", ticker.String()),
+				zap.String("ticker", target.String()),
 				zap.Any("conversions", path),
 			)
 
 			continue
 		}
 
-		convertedPrices = append(convertedPrices, convertedPrice)
+		convertedPrices = append(convertedPrices, adjustedPrice)
+		m.logger.Debug(
+			"calculated converted price",
+			zap.String("ticker", target.String()),
+			zap.String("price", adjustedPrice.String()),
+			zap.Any("conversions", path.Operations),
+		)
 	}
 
 	return convertedPrices
 }
 
-// CalculateConvertedPrice converts a set of median prices to a target ticker using a set of
-// conversion operations.
-func (m *MedianAggregator) CalculateConvertedPrice(
+// CalculateAdjustedPrice calculates an adjusted price for a given set of operations (if applicable).
+// In particular, this assumes that every operation is either:
+//
+//  1. A direct conversion from the base ticker to the target ticker i.e. we want BTC/USD and
+//     we have BTC/USD from a provider (e.g. Coinbase).
+//  2. We need to convert the price of a given asset against the index price of an asset.
+//
+// In the first case, we can simply return the price of the provider. In the second case, we need
+// to adjust the price by the index price of the asset. If the index price is not available, we
+// return an error.
+func (m *MedianAggregator) CalculateAdjustedPrice(
 	target mmtypes.Ticker,
-	path mmtypes.Path,
-	medians types.TickerPrices,
+	operations []mmtypes.Operation,
 ) (*big.Int, error) {
-	if err := path.ValidateBasic(); err != nil {
+	price, err := m.GetProviderPrice(operations[0])
+	if err != nil {
 		return nil, err
 	}
 
-	// Ensure that the conversion is valid.
-	if !path.Match(target.String()) {
-		return nil, fmt.Errorf("path does not match target %s: %s", target.String(), path.String())
+	// If we have a single operation, then we can simply return the price. This implies that
+	// we have a direct conversion from the base ticker to the target ticker.
+	if len(operations) == 1 {
```

Comment:
we should ensure that:

```
foreach path
     len(path.operations) == 1 OR == 2

```

Reply:
We have this check in the parent function that calls this.

Reply:
modified so that this is verified in the same func